File: python/services/vectorindex.py
"""In-memory semantic index.

Chunks are embedded with the embedding model reached over the currently selected
mode (direct / APIM / AI Gateway). If no embedding model is configured the index
falls back to a deterministic hashed TF-IDF vector so the demo still works
end to end - the UI reports which of the two produced the vectors.
"""
from __future__ import annotations

import math
import re
import uuid
import zlib
from dataclasses import dataclass, field
from threading import Lock

HASH_DIMS = 512
# Hyphen and underscore are separators, not word characters: a document saying
# "Operation-Location" has to match a question asking about "operation location".
_TOKEN = re.compile(r"[a-z0-9][a-z0-9']*")


# ── chunking ──────────────────────────────────────────────────────────
def chunk_text(text: str, size: int = 1200, overlap: int = 150) -> list[str]:
    """Split on paragraph boundaries, packing up to `size` characters per chunk."""
    text = re.sub(r"\n{3,}", "\n\n", (text or "").strip())
    if not text:
        return []

    paragraphs = [p.strip() for p in text.split("\n\n") if p.strip()]
    chunks: list[str] = []
    current = ""

    for para in paragraphs:
        # A single oversized paragraph is hard-split.
        while len(para) > size:
            if current:
                chunks.append(current)
                current = ""
            cut = para.rfind(" ", 0, size)
            cut = cut if cut > size // 2 else size
            chunks.append(para[:cut].strip())
            para = para[cut:].lstrip()

        if not current:
            current = para
        elif len(current) + len(para) + 2 <= size:
            current = f"{current}\n\n{para}"
        else:
            chunks.append(current)
            tail = current[-overlap:] if overlap else ""
            current = f"{tail}\n\n{para}".strip() if tail else para

    if current:
        chunks.append(current)
    return chunks


# ── vectors ───────────────────────────────────────────────────────────
def hashed_vector(text: str, dims: int = HASH_DIMS) -> list[float]:
    """Deterministic hashed TF vector with sub-linear term weighting, L2 normalised."""
    counts: dict[int, float] = {}
    for token in _TOKEN.findall(text.lower()):
        if len(token) < 2:
            continue
        # crc32 rather than hash(): stable across processes and restarts.
        slot = zlib.crc32(token.encode("utf-8")) % dims
        counts[slot] = counts.get(slot, 0.0) + 1.0

    vec = [0.0] * dims
    for slot, count in counts.items():
        vec[slot] = 1.0 + math.log(count)

    norm = math.sqrt(sum(v * v for v in vec))
    return [v / norm for v in vec] if norm else vec


def cosine(a: list[float], b: list[float]) -> float:
    if not a or not b or len(a) != len(b):
        return 0.0
    dot = na = nb = 0.0
    for x, y in zip(a, b):
        dot += x * y
        na += x * x
        nb += y * y
    if na == 0 or nb == 0:
        return 0.0
    return dot / math.sqrt(na * nb)


# ── index ─────────────────────────────────────────────────────────────
@dataclass
class Chunk:
    doc_id: str
    doc_name: str
    ordinal: int
    text: str
    vector: list[float]


@dataclass
class Document:
    id: str
    name: str
    size_bytes: int
    extracted_via: str
    indexed_via: str
    pages: int | None = None
    elapsed_ms: int = 0
    chunks: list[Chunk] = field(default_factory=list)

    def to_json(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "sizeBytes": self.size_bytes,
            "chunks": len(self.chunks),
            "pages": self.pages,
            "extractedVia": self.extracted_via,
            "indexedVia": self.indexed_via,
            "elapsedMs": self.elapsed_ms,
        }


class VectorIndex:
    """Process-lifetime store. Nothing is persisted - restart clears the index."""

    def __init__(self) -> None:
        self._docs: dict[str, Document] = {}
        self._lock = Lock()

    def add(self, name: str, size_bytes: int, extracted_via: str, indexed_via: str,
            pages: int | None, texts: list[str], vectors: list[list[float]],
            elapsed_ms: int) -> Document:
        doc = Document(
            id=uuid.uuid4().hex[:12], name=name, size_bytes=size_bytes,
            extracted_via=extracted_via, indexed_via=indexed_via,
            pages=pages, elapsed_ms=elapsed_ms,
        )
        doc.chunks = [
            Chunk(doc.id, name, i, t, v)
            for i, (t, v) in enumerate(zip(texts, vectors))
        ]
        with self._lock:
            self._docs[doc.id] = doc
        return doc

    def remove(self, doc_id: str) -> bool:
        with self._lock:
            return self._docs.pop(doc_id, None) is not None

    def documents(self) -> list[Document]:
        with self._lock:
            return list(self._docs.values())

    def total_chunks(self) -> int:
        return sum(len(d.chunks) for d in self.documents())

    def is_empty(self) -> bool:
        return self.total_chunks() == 0

    def search(self, query_vector: list[float], top_k: int = 4,
               min_score: float = 0.05) -> list[tuple[Chunk, float]]:
        scored = [
            (chunk, cosine(query_vector, chunk.vector))
            for doc in self.documents()
            for chunk in doc.chunks
        ]
        scored.sort(key=lambda pair: pair[1], reverse=True)
        return [(c, s) for c, s in scored[:top_k] if s >= min_score]


index = VectorIndex()

File: dotnet/AiGatewayDemo/Services/VectorIndex.cs
using System.Collections.Concurrent;
using System.Text;
using System.Text.RegularExpressions;

namespace AiGatewayDemo.Services;

public sealed record Chunk(string DocId, string DocName, int Ordinal, string Text, float[] Vector);

public sealed class IndexedDocument
{
    public required string Id { get; init; }
    public required string Name { get; init; }
    public required long SizeBytes { get; init; }
    public required string ExtractedVia { get; init; }
    public required string IndexedVia { get; init; }
    public int? Pages { get; init; }
    public int ElapsedMs { get; set; }
    public List<Chunk> Chunks { get; } = [];

    public object ToJson() => new
    {
        id = Id,
        name = Name,
        sizeBytes = SizeBytes,
        chunks = Chunks.Count,
        pages = Pages,
        extractedVia = ExtractedVia,
        indexedVia = IndexedVia,
        elapsedMs = ElapsedMs,
    };
}

/// <summary>
/// In-memory semantic index. Chunks are embedded over the selected mode; when no
/// embedding model is reachable the hashed term-frequency fallback keeps the demo
/// working end to end, and the UI reports which one produced the vectors.
/// Process lifetime only - a restart clears it.
/// </summary>
public sealed partial class VectorIndex
{
    public const int HashDims = 512;

    // Hyphen and underscore are separators, not word characters: a document saying
    // "Operation-Location" has to match a question asking about "operation location".
    [GeneratedRegex("[a-z0-9][a-z0-9']*", RegexOptions.Compiled)]
    private static partial Regex TokenRegex();

    [GeneratedRegex("\n{3,}", RegexOptions.Compiled)]
    private static partial Regex BlankRunRegex();

    private readonly ConcurrentDictionary<string, IndexedDocument> _docs = new();

    // ── chunking ──────────────────────────────────────────────────────
    public static List<string> ChunkText(string text, int size = 1200, int overlap = 150)
    {
        var chunks = new List<string>();
        text = BlankRunRegex().Replace((text ?? "").Replace("\r\n", "\n").Trim(), "\n\n");
        if (text.Length == 0) return chunks;

        var paragraphs = text.Split("\n\n", StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
        var current = new StringBuilder();

        foreach (var original in paragraphs)
        {
            var para = original;

            // A single oversized paragraph is hard-split.
            while (para.Length > size)
            {
                if (current.Length > 0)
                {
                    chunks.Add(current.ToString());
                    current.Clear();
                }

                var cut = para.LastIndexOf(' ', Math.Min(size, para.Length - 1));
                if (cut <= size / 2) cut = size;
                chunks.Add(para[..cut].Trim());
                para = para[cut..].TrimStart();
            }

            if (para.Length == 0) continue;

            if (current.Length == 0)
            {
                current.Append(para);
            }
            else if (current.Length + para.Length + 2 <= size)
            {
                current.Append("\n\n").Append(para);
            }
            else
            {
                var finished = current.ToString();
                chunks.Add(finished);
                current.Clear();
                var tail = overlap > 0 && finished.Length > overlap ? finished[^overlap..] : "";
                if (tail.Length > 0) current.Append(tail).Append("\n\n");
                current.Append(para);
            }
        }

        if (current.Length > 0) chunks.Add(current.ToString());
        return chunks;
    }

    // ── vectors ───────────────────────────────────────────────────────
    /// <summary>Deterministic hashed term-frequency vector, L2 normalised.</summary>
    public static float[] HashedVector(string text, int dims = HashDims)
    {
        var counts = new Dictionary<int, float>();
        foreach (Match match in TokenRegex().Matches((text ?? "").ToLowerInvariant()))
        {
            if (match.Length < 2) continue;
            var slot = (int)(StableHash(match.ValueSpan) % (uint)dims);
            counts[slot] = counts.GetValueOrDefault(slot) + 1f;
        }

        var vector = new float[dims];
        foreach (var (slot, count) in counts)
            vector[slot] = 1f + MathF.Log(count);

        var norm = MathF.Sqrt(vector.Sum(v => v * v));
        if (norm > 0)
            for (var i = 0; i < dims; i++) vector[i] /= norm;

        return vector;
    }

    /// <summary>FNV-1a: stable across processes, unlike string.GetHashCode.</summary>
    private static uint StableHash(ReadOnlySpan<char> text)
    {
        uint hash = 2166136261;
        foreach (var c in text)
        {
            hash = (hash ^ (byte)c) * 16777619;
            hash = (hash ^ (byte)(c >> 8)) * 16777619;
        }
        return hash;
    }

    public static double Cosine(float[] a, float[] b)
    {
        if (a.Length == 0 || a.Length != b.Length) return 0;
        double dot = 0, na = 0, nb = 0;
        for (var i = 0; i < a.Length; i++)
        {
            dot += a[i] * b[i];
            na += a[i] * a[i];
            nb += b[i] * b[i];
        }
        return na == 0 || nb == 0 ? 0 : dot / Math.Sqrt(na * nb);
    }

    // ── index ─────────────────────────────────────────────────────────
    public IndexedDocument Add(string name, long sizeBytes, string extractedVia, string indexedVia,
        int? pages, IReadOnlyList<string> texts, IReadOnlyList<float[]> vectors, int elapsedMs)
    {
        var doc = new IndexedDocument
        {
            Id = Guid.NewGuid().ToString("n")[..12],
            Name = name,
            SizeBytes = sizeBytes,
            ExtractedVia = extractedVia,
            IndexedVia = indexedVia,
            Pages = pages,
            ElapsedMs = elapsedMs,
        };

        for (var i = 0; i < texts.Count; i++)
            doc.Chunks.Add(new Chunk(doc.Id, name, i, texts[i], vectors[i]));

        _docs[doc.Id] = doc;
        return doc;
    }

    public bool Remove(string id) => _docs.TryRemove(id, out _);

    public IReadOnlyList<IndexedDocument> Documents() => _docs.Values.ToList();

    public int TotalChunks() => _docs.Values.Sum(d => d.Chunks.Count);

    public bool IsEmpty => TotalChunks() == 0;

    public List<(Chunk Chunk, double Score)> Search(float[] queryVector, int topK = 4, double minScore = 0.05)
        => _docs.Values
            .SelectMany(d => d.Chunks)
            .Select(c => (Chunk: c, Score: Cosine(queryVector, c.Vector)))
            .OrderByDescending(x => x.Score)
            .Take(topK)
            .Where(x => x.Score >= minScore)
            .ToList();
}
